src/NLog/LogEventInfo.cs

```diff
@@ -281,16 +296,47 @@ public string FormattedMessage
         /// <summary>
         /// Gets the dictionary of per-event context properties.
         /// </summary>
-        public IDictionary<object, object> Properties
+        public IDictionary<object,object> Properties { get { return this.properties ?? (this.properties = new PropertiesDictionary()); } }
+
+        /// <summary>
+        /// Gets the named parameters extracted from parsing <see cref="Message"/> as MessageTemplate
+        /// </summary>
+        public IEnumerable<MessageTemplateParameter> MessageTemplateParameters
         {
             get
             {
-                if (this.properties == null)
+                if (this.properties != null && this.properties.MessageProperties != null && this.properties.MessageProperties.Count > 0)
                 {
-                    this.properties = new Dictionary<object, object>();
+                    return this.properties.MessageProperties;
+                }
+                else if (this.parameters != null && this.parameters.Length > 0)
+                {
+                    var messageParameters = new MessageTemplateParameter[parameters.Length];
+                    for (int i = 0; i < parameters.Length; ++i)
+                    {
+                        string parameterName;
+                        switch (i)
+                        {
+                            case 0: parameterName = "0"; break;
+                            case 1: parameterName = "1"; break;
+                            case 2: parameterName = "2"; break;
+                            case 3: parameterName = "3"; break;
+                            case 4: parameterName = "4"; break;
+                            case 5: parameterName = "5"; break;
+                            case 6: parameterName = "6"; break;
+                            case 7: parameterName = "7"; break;
+                            case 8: parameterName = "8"; break;
+                            case 9: parameterName = "9"; break;
+                            default: parameterName = i.ToString(); break;
+                        }
+                        messageParameters[i] = new MessageTemplateParameter(parameterName, parameters[i], null);
```

Comment:
would not it been better to use `yield return` here? (depends on the return interface)

Reply:
I like for-loops better, when not needing the state-machine.